<a href="https://colab.research.google.com/github/innovinitylabs/learning-pytorch/blob/main/00_pytorch_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [231]:
import torch

In [232]:
# mps_device = torch.device('mps')

In [233]:
# class Conf: dev = torch.device("mps")

# introduction to tensors

## creating tensors

pytorch tensors are created using `torch.tensors`

### Scalar

In [234]:
scalar = torch.tensor(7)
scalar

tensor(7)

to get dimensions

In [235]:
scalar.ndim

0

to get the number without type

In [236]:
scalar.item()

7

### Vector

In [237]:
vector = torch.tensor([7, 7])
vector

tensor([7, 7])

In [238]:
vector.ndim

1

easy way to remember dimension is by number of square brackets

In [239]:
vector.shape

torch.Size([2])

shape gives how elements in tensor are arranged.

### MATRIX

In [240]:
MATRIX = torch.tensor([[7, 8],
                       [9,10]])
MATRIX

tensor([[ 7,  8],
        [ 9, 10]])

In [241]:
MATRIX.ndim

2

In [242]:
MATRIX.shape

torch.Size([2, 2])

In [243]:
MATRIX[0]

tensor([7, 8])

### TENSOR

In [244]:
TENSOR = torch.tensor([[[1, 2, 3],
                        [3, 6, 9],
                        [2, 4, 5]]])
TENSOR

tensor([[[1, 2, 3],
         [3, 6, 9],
         [2, 4, 5]]])

In [245]:
TENSOR.ndim

3

In [246]:
TENSOR.shape

torch.Size([1, 3, 3])

In [247]:
TENSOR[0]


tensor([[1, 2, 3],
        [3, 6, 9],
        [2, 4, 5]])

### Random Tensors

random tensors are important because many Neural networks starts with random numbers

In [248]:
random_tensor = torch.rand(3, 4)
random_tensor

tensor([[0.0183, 0.5854, 0.6620, 0.0158],
        [0.5309, 0.9056, 0.6011, 0.6072],
        [0.5147, 0.7654, 0.5434, 0.3774]])

In [249]:
random_tensor.ndim

2

In [250]:
random_huge_tensor = torch.zeros(4,5)
random_huge_tensor

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])

In [251]:
random_huge_tensor.dtype

torch.float32

In [252]:
torch.__version__

'2.1.0'

In [253]:
zttendep = torch.arange(0, 10)
zttendep

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [254]:
ten = torch.tensor([11, 22, 33])

In [255]:
ten * 10

tensor([110, 220, 330])

In [256]:
torch.mul(ten, 11)

tensor([121, 242, 363])

In [257]:
ten * ten

tensor([ 121,  484, 1089])

In [258]:
twotwo = torch.tensor([[1, 2],
                       [3, 4]])

In [259]:
twotwo * twotwo

tensor([[ 1,  4],
        [ 9, 16]])

In [260]:
torch.matmul(twotwo, twotwo)

tensor([[ 7, 10],
        [15, 22]])

In [261]:
torch.matmul(ten, ten)

tensor(1694)

In [262]:
tensor1 = torch.randn(10, 3, 4)
tensor2 = torch.randn(4)
torch.matmul(tensor1, tensor2).size()

torch.Size([10, 3])

In [263]:
tensor1.size()

torch.Size([10, 3, 4])

In [264]:
tensor2.size()

torch.Size([4])

In [265]:
tensor1, tensor2

(tensor([[[ 0.8256,  0.8756,  0.2960,  0.6394],
          [ 1.2406,  0.7299, -0.8683,  0.6726],
          [ 0.2188, -1.2224, -0.9322, -0.3832]],
 
         [[ 1.4027,  1.9010, -1.1275,  0.7207],
          [-0.8487,  1.8939, -0.0325,  0.3191],
          [-0.5700,  0.9400,  1.5548, -0.3399]],
 
         [[ 1.6272, -0.2411, -2.1021, -0.0179],
          [ 0.2241,  1.8586, -1.3166,  1.6165],
          [-1.6016,  0.4660, -0.1961,  0.3063]],
 
         [[-1.8420,  0.8163, -0.4833,  0.0640],
          [ 1.3794,  1.1527,  0.0762,  0.0557],
          [-0.6163, -1.8684, -0.0034, -1.0512]],
 
         [[ 0.0517,  0.7329, -0.7448,  2.2900],
          [ 0.1214,  0.4600,  0.1848,  0.4180],
          [ 1.5207,  0.5975,  0.5968, -0.5111]],
 
         [[-0.2617, -1.3285,  0.6058,  1.5191],
          [-0.2140, -1.0078,  0.8051,  0.7826],
          [-0.8145, -1.4938, -0.5081,  0.2437]],
 
         [[-0.3933,  0.2791,  0.8211,  0.6328],
          [ 0.8020, -0.4726, -0.1192,  0.4446],
          [ 0.9207, -0

In [266]:
torch.matmul(tensor1, tensor2)

tensor([[-3.1603, -2.8748,  2.9114],
        [-5.2247, -3.3230, -1.4621],
        [-0.1308, -5.1787,  0.0810],
        [-0.0331, -3.3911,  5.2804],
        [-3.8339, -1.5171, -1.9530],
        [ 0.8253,  0.9516,  3.3952],
        [-1.2122, -0.1965, -1.4294],
        [ 2.5041, -0.1711,  0.5837],
        [ 2.0377,  2.9412,  2.7466],
        [-4.9906, -1.5184, -1.0068]])

In [267]:
torch.matmul(torch.rand(3,2),torch.rand(2,3))

tensor([[0.3304, 0.2364, 0.3512],
        [0.1830, 0.1345, 0.1965],
        [0.6208, 0.5125, 0.6981]])

In [268]:
x = torch.arange(0, 100, 10)
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [269]:
x.argmax()

tensor(9)

In [270]:
x.reshape(5,2)

tensor([[ 0, 10],
        [20, 30],
        [40, 50],
        [60, 70],
        [80, 90]])

In [271]:
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [272]:
torch.reshape(x, (5,2))

tensor([[ 0, 10],
        [20, 30],
        [40, 50],
        [60, 70],
        [80, 90]])

In [273]:
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [274]:
torch.stack([x,x,x], dim= 0)

tensor([[ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90],
        [ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90],
        [ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90]])

In [275]:
torch.stack([x,x,x], dim= 1)

tensor([[ 0,  0,  0],
        [10, 10, 10],
        [20, 20, 20],
        [30, 30, 30],
        [40, 40, 40],
        [50, 50, 50],
        [60, 60, 60],
        [70, 70, 70],
        [80, 80, 80],
        [90, 90, 90]])

In [276]:
x = torch.zeros(2, 1, 2, 1, 2)
x

tensor([[[[[0., 0.]],

          [[0., 0.]]]],



        [[[[0., 0.]],

          [[0., 0.]]]]])

In [277]:
y = torch.squeeze(x)
y

tensor([[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]])

In [278]:
x[:, 0] = 1
x , y

(tensor([[[[[1., 1.]],
 
           [[1., 1.]]]],
 
 
 
         [[[[1., 1.]],
 
           [[1., 1.]]]]]),
 tensor([[[1., 1.],
          [1., 1.]],
 
         [[1., 1.],
          [1., 1.]]]))

In [279]:
z = torch.unsqueeze(y, dim=1)
z

tensor([[[[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]]]])

In [280]:
torch.unsqueeze(y, dim=0)

tensor([[[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]]])

In [281]:
torch.unsqueeze(y, dim=1)

tensor([[[[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]]]])

In [282]:
torch.unsqueeze(y, dim=3)

tensor([[[[1.],
          [1.]],

         [[1.],
          [1.]]],


        [[[1.],
          [1.]],

         [[1.],
          [1.]]]])

#### to check if squeeze shares memory

In [283]:
sqz = torch.arange(0,12, 2)
sqz

tensor([ 0,  2,  4,  6,  8, 10])

In [284]:
sqz = sqz.unsqueeze(dim=0)

In [285]:
sqz

tensor([[ 0,  2,  4,  6,  8, 10]])

In [286]:
sqz2 = sqz.squeeze( dim = 0)
sqz2

tensor([ 0,  2,  4,  6,  8, 10])

In [287]:
sqz2[1] = 69
sqz2

tensor([ 0, 69,  4,  6,  8, 10])

In [288]:
sqz

tensor([[ 0, 69,  4,  6,  8, 10]])

Reproducablity. random seed

In [289]:
r_seed = 42
torch.manual_seed(r_seed)
rand_a = torch.rand(3,2)
rand_a

tensor([[0.8823, 0.9150],
        [0.3829, 0.9593],
        [0.3904, 0.6009]])

In [290]:
torch.rand(3,2)

tensor([[0.2566, 0.7936],
        [0.9408, 0.1332],
        [0.9346, 0.5936]])

In [291]:
torch.manual_seed(42)
torch.rand(3,2)

tensor([[0.8823, 0.9150],
        [0.3829, 0.9593],
        [0.3904, 0.6009]])

In [292]:
torch.manual_seed(42)
rand_a == torch.rand(3,2)

tensor([[True, True],
        [True, True],
        [True, True]])

In [293]:
rand_a.device

device(type='cpu')

In [294]:
# mps_rans = rand_a.to(mps_device)

rand_a still stays on CPU. have to assign it to a variable to hold the migrated tensor on `mps` like above

In [295]:
# mps_rans.device

In [296]:
!nvidia-smi

zsh:1: command not found: nvidia-smi


# Practice

using template given in course for ease

In [297]:
import torch
seed = 42

In [298]:
print(f'without setting any prior manual seed \n{torch.rand(3,3)}') #RANDOM

without setting any prior manual seed 
tensor([[0.2566, 0.7936, 0.9408],
        [0.1332, 0.9346, 0.5936],
        [0.8694, 0.5677, 0.7411]])


### torch.`random`.manual_seed()

In [299]:
torch.random.manual_seed(seed)
print(f'with random manual seed \n{torch.rand(3,3)}\n') #
print(f'call 1 \n{torch.rand(3,3)}')
print(f'call 2 \n{torch.rand(3,3)}')

with random manual seed 
tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009],
        [0.2566, 0.7936, 0.9408]])

call 1 
tensor([[0.1332, 0.9346, 0.5936],
        [0.8694, 0.5677, 0.7411],
        [0.4294, 0.8854, 0.5739]])
call 2 
tensor([[0.2666, 0.6274, 0.2696],
        [0.4414, 0.2969, 0.8317],
        [0.1053, 0.2695, 0.3588]])


### torch.manual_seed()

In [300]:
# torch.manual_seed(seed)
print(f'call 3 without setting seed above, call 1 if set \n{torch.rand(3,3)} \n') #
print(f'call 4 without setting seed above, call 2 if set \n{torch.rand(3,3)}') #
print(f'call 4 without setting seed above, call 3 if set \n{torch.rand(3,3)}')

call 3 without setting seed above, call 1 if set 
tensor([[0.1994, 0.5472, 0.0062],
        [0.9516, 0.0753, 0.8860],
        [0.5832, 0.3376, 0.8090]]) 

call 4 without setting seed above, call 2 if set 
tensor([[0.5779, 0.9040, 0.5547],
        [0.3423, 0.6343, 0.3644],
        [0.7104, 0.9464, 0.7890]])
call 4 without setting seed above, call 3 if set 
tensor([[0.2814, 0.7886, 0.5895],
        [0.7539, 0.1952, 0.0050],
        [0.3068, 0.1165, 0.9103]])


In [301]:
print(f'call n \n{torch.rand(3,3)}') #totally random because it changes every time its called

call n 
tensor([[0.6440, 0.7071, 0.6581],
        [0.4913, 0.8913, 0.1447],
        [0.5315, 0.1587, 0.6542]])


In [302]:
torch.manual_seed(69)
torch.rand(3,3)

tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])

In [303]:
torch.rand(3,3)

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])

In [304]:
torch.manual_seed(69)
torch.rand(3,3)

tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])

In [305]:
torch.rand(3,3)

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])

In [306]:
torch.manual_seed(69)
torch.rand(3,3)
torch.rand(3,3)
torch.rand(3,3)



tensor([[0.6444, 0.0099, 0.5925],
        [0.9631, 0.6958, 0.9157],
        [0.5523, 0.2344, 0.3262]])

##### to reset the manual seed test

In [307]:
initialseed = torch.random.get_rng_state()

In [308]:
torch.manual_seed(69)
print(torch.rand(3,3))



tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])


In [309]:
print(torch.rand(3,3))

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])


In [310]:
# print(torch.rand(3,3))

torch.random.set

In [311]:
torch.random.set_rng_state(initialseed)

In [312]:
print(torch.rand(3,3))

tensor([[0.4521, 0.7020, 0.6274],
        [0.0945, 0.8525, 0.3572],
        [0.7492, 0.3579, 0.5453]])



# print(torch.rand(3,3)) 3rd call should be
account_circle
tensor([[0.6444, 0.0099, 0.5925],
        [0.9631, 0.6958, 0.9157],
        [0.5523, 0.2344, 0.3262]])

# To reset manual seed

has to get state of the seed and save it to variable and set back.

In [313]:
initialseed = torch.random.get_rng_state()

saving initial state

In [314]:
torch.manual_seed(69)
print(f'call 1 \n{torch.rand(1,3)}')
print(f'call 2 \n{torch.rand(1,3)}')
print(f'call 3 \n{torch.rand(1,3)}')

call 1 
tensor([[0.8398, 0.8042, 0.1213]])
call 2 
tensor([[0.5309, 0.6646, 0.4077]])
call 3 
tensor([[0.0888, 0.2429, 0.7053]])


In [315]:
torch.manual_seed(69)
print(f'call 1 \n{torch.rand(1,3)}')
print(f'call 2 \n{torch.rand(1,3)}')

call 1 
tensor([[0.8398, 0.8042, 0.1213]])
call 2 
tensor([[0.5309, 0.6646, 0.4077]])


In [316]:
torch.random.set_rng_state(initialseed)

reset seed state

In [317]:
print(f'call 3 \n{torch.rand(1,3)}')

call 3 
tensor([[0.8171, 0.0570, 0.4560]])
